src/dscanner/analysis/cyclomatic_complexity.d

```diff
@@ -0,0 +1,260 @@
+// Distributed under the Boost Software License, Version 1.0.
+//    (See accompanying file LICENSE_1_0.txt or copy at
+//          http://www.boost.org/LICENSE_1_0.txt)
+
+module dscanner.analysis.cyclomatic_complexity;
+
+import dparse.ast;
+import dparse.lexer;
+import dsymbol.scope_ : Scope;
+import dscanner.analysis.base;
+import dscanner.analysis.helpers;
+
+import std.format;
+
+/// Implements a basic cyclomatic complexity algorithm using the AST.
+///
+/// Issues a warning on functions whenever the cyclomatic complexity of them
+/// passed over a configurable threshold.
+///
+/// The complexity score starts at 1 and is increased each time on
+/// - `if`
+/// - switch `case`
+/// - any loop
+/// - `&&`
+/// - `||`
+/// - `?:` (ternary operator)
+/// - `throw`
+/// - `catch`
+/// - `return`
+/// - `break` (unless in case)
+/// - `continue`
+/// - `goto`
+/// - function literals
+///
+/// See: https://en.wikipedia.org/wiki/Cyclomatic_complexity
+final class CyclomaticComplexityCheck : BaseAnalyzer
+{
+	/// Message key emitted when the threshold is reached
+	enum string KEY = "dscanner.metric.cyclomatic_complexity";
+	/// Human readable message emitted when the threshold is reached
+	enum string MESSAGE = "Cyclomatic complexity of this function is %s.";
+	mixin AnalyzerInfo!"cyclomatic_complexity";
+
+	/// Maximum cyclomatic complexity. Once the cyclomatic complexity is greater
+	/// than this threshold, a warning is issued.
+	///
+	/// By default 50 is used as threshold, which is considered almost
+	/// unmaintainable / untestable.
+	///
+	/// For clean development a threshold like 20 can be used instead.
+	int maxCyclomaticComplexity;
+
+	///
+	this(string fileName, const(Scope)* sc, bool skipTests = false,
+		int maxCyclomaticComplexity = 50)
+	{
+		super(fileName, sc, skipTests);
+		this.maxCyclomaticComplexity = maxCyclomaticComplexity;
+	}
+
+	// based on http://cyvis.sourceforge.net/cyclomatic_complexity.html
```

Comment:
:+1: Maybe put the source of the definition used here somewhere more prominent like the module header.